// @iota/wallet/out/lib/bindings.js
"use strict";
// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.migrateStrongholdSnapshotV2ToV3 = exports.destroy = exports.listenWallet = exports.messageHandlerNew = exports.sendMessageAsync = exports.internalInitLogger = void 0;
// @ts-ignore: path is set to match runtime transpiled js path
const addon = require("../../build/Release/index.node");
const { initLogger, sendMessage, messageHandlerNew, listen, destroy, migrateStrongholdSnapshotV2ToV3, } = addon;
exports.internalInitLogger = initLogger;
exports.messageHandlerNew = messageHandlerNew;
exports.destroy = destroy;
exports.migrateStrongholdSnapshotV2ToV3 = migrateStrongholdSnapshotV2ToV3;
const listenWallet = (eventTypes, callback, handler) => {
    listen(eventTypes, callback, handler);
    return Promise.resolve();
};
exports.listenWallet = listenWallet;
const sendMessageAsync = (message, handler) => new Promise((resolve, reject) => {
    sendMessage(message, handler, (error, result) => {
        if (error) {
            reject(error);
        }
        else {
            resolve(result);
        }
    });
});
exports.sendMessageAsync = sendMessageAsync;
